cmd/agent/app/agent.go

```diff
@@ -15,21 +28,96 @@ func NewAgentCommand() *cobra.Command {
 		Use:  "Agent",
 		Long: "Agent client run in user cluster",
 		RunE: func(cmd *cobra.Command, args []string) error {
+			o.Print()
 			if err := o.Validate(); err != nil {
 				return err
 			}
 
-			client, err := agent.NewAgent(o.AgentOptions)
+			stopCh := signals.SetupSignalHandler()
+			agentrunfunc := func(ctx context.Context) {
+				client, err := agent.NewAgent(o.AgentOptions)
+				if err != nil {
+					klog.Fatal(err)
+				}
+
+				if err := client.Run(); err != nil {
+					klog.Fatal(err)
+				}
+
+				if err := client.Wait(); err != nil {
+					klog.Fatal()
+				}
+
+				select {}
+
+			}
+
+			ctx, cancel := context.WithCancel(context.Background())
+			defer cancel()
+
+			go func() {
+				<-stopCh
+				cancel()
+			}()
+
+			if !o.LeaderElect {
+				agentrunfunc(ctx)
+				return nil
+			}
+
+			id, err := os.Hostname()
+			if err != nil {
+				return err
+			}
+
+			config, err := clientcmd.BuildConfigFromFlags("", o.AgentOptions.Kubeconfig)
 			if err != nil {
-				klog.Error(err)
+				klog.Errorf("Failed to create config from kubeconfig file, %v", err)
 				return err
 			}
 
-			if err := client.Run(); err != nil {
-				klog.Error(err)
+			kubernetesClient, err := kubernetes.NewForConfig(config)
+			if err != nil {
+				klog.Errorf("Failed to create kubernetes clientSets, %v", err)
 				return err
 			}
-			return client.Wait()
+			// add a uniquifier so that two processes on the same host don't accidentally both become active
+			id = id + "_" + string(uuid.NewUUID())
+
+			// TODO: change lockType to lease
```

Comment:
Why do we still have those comments? The lockType is already Lease

Reply:
It was copied from cmd/agent/app/proxy.go. I'm not sure if k8s v1.15 supports the lease lock type.